lessons 3/4: parsing and disambiguation

dwightguth opened the pull request.

Commit:
lesson 3: parsing

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+string represents a sentence in the formal language defined by the above
+grammar. K then parses it automatically into an AST of that term, and then,
+since the symbol `colorOf` is a function, it evaluates that function using
+the rules that define it.
```
```suggestion
parsed and interpreted by K's interpreter. K parses this string into an AST of the term, then evaluates the `colorOf` function using the functional rules provided in the definition.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+represents the AST of the program that you just parsed. You ought to be able
+to recognize the basic shape of it by seeing the words `true`, `false`, and
+`And` in there. This is `KORE`, the intermediate representation of K, and we
+will cover it in detail later.
```
```suggestion
`And` in there. This is **Kore**, the intermediate representation of K, and we
will cover it in detail later.
```

ehildenb replied:
You may also want to consider showing `--output kast` here, because later in the lesson the parsing ambiguities are given in Kast form.

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+able to be parsed into more than one distinct AST. In this example, it can't
+decide whether it should be parsed as `(true && false) || false` or as
+`true && (false || false)`. As a result, it reports the error to the user.
+
```
```suggestion
This error is saying that `kast` was unable to parse this
program because it is ambiguous. K's just-in-time parser is a GLL parser, which means it can handle the full generality
of context-free grammars, including grammars which are ambiguous. An
ambiguous grammar is one where the same string (program) can be validly parsed as multiple distinct ASTs. In this example, it can't
decide whether it should be parsed as `(true && false) || false` or as
`true && (false || false)`. As a result, it reports the error to the user.

```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+productions with the bracket attribute are not represented in the parsed AST
+of a term, and the child of the bracket is folded immediately into the parent
+term. This is the reason for the requirement that a bracket production have
+a single non-terminal of the same sort as the production itself.
```
```suggestion
If you parse these programs with `kast`, you will once again get a single
unique AST with no error. If you look, you will notice that the bracket itself
does not appear in the AST. In fact, this is a property unique to brackets:
productions with the bracket attribute are not represented in the parsed AST
of a term, and the child of the bracket is folded immediately into the parent
term. This is the reason for the requirement that a bracket production have
a single non-terminal of the same sort as the production itself.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+languages. Programming language designers instead prefer to use a notion of
+operator precedence and associativity to make expression grammars unambiguous
+by automatically choosing certain parses over other parses in cases where the
+grammar would otherwise be ambiguous.
```
```suggestion
Of course, in practice, very few formal languages outside the domain of
natural language processing are ambiguous. The main reason for this is that
parsing unambiguous languages is asymptotically faster than parsing ambiguous
languages. Programming language designers use the notions of
operator precedence and associativity to make expression grammars unambiguous.
These tools work by instructing the parser which ASTs to select in the case of ambiguities; often times it's possible to remove _all_ the ambiguities in a grammar with these techniques.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+inextricably linked, this is generally discouraged. Instead, we use the
+approach of explicitly expressing the relative precedence of different
+operators in different situations in order to resolve the ambiguity.
+
```
This block has the tone of "contradicting" the previous block of text, even though it agrees with it. Doesn't seem to add much either, so I would just remove it.

```suggestion
```

dwightguth replied:
This paragraph actually does serve an important role. It explains to anyone who is previously familiar with other parser generators an important way in which parser generation in K is usually done differently than they might be used to.

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+
+For example, in C, `&&` binds tighter in precedence than `||`, meaning that
+the expression `true && false || false` would be parsed as
+`(true && false) || false`.
```
```suggestion
For example, in C, `&&` binds tighter in precedence than `||`, meaning that
the expression `true && false || false` has only one valid parse:
`(true && false) || false`.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+appear as the **direct child** of a symbol with a **greater priority** (i.e.,
+one that **binds tighter**. In this case, the `>` operator can be seen as a
+**less-than** operator describing a transitive partial ordering on the 
+productions in the production block, expressing their relative priority.
```
```suggestion
one that **binds tighter**. In this case, the `>` operator can be seen as a
**greater-than** operator describing a transitive partial ordering on the 
productions in the production block, expressing their relative priority.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+successfully because the `||` operator is no longer the **direct** child of the
+`&&` operator, but is instead the direct child of the `()` operator, and the
+`&&` operator is an **indirect** parent. Indirect parents are completely
+ignored when disambiguating; only the direct parent/child relation matters.
```
```suggestion
To see this more concretely, let's look again at the program
`true && false || false`. As noted before, previously this program was
ambiguous because the parser could either choose the `&&` to be a child of the `||`
or vice versa. However, because a symbol with lesser priority (in this case `||`)
cannot appear as the direct child of a symbol with greater priority (in this case `&&`),
the parser can now **reject** the parse where `||` is under the `&&` operator. As a
result, we are left with the unambiguous parse `(true && false) || false`.
Similarly, `true || false && false` parses unambiguously as
`true || (false && false)`. Conversely, if the user explicitly wants the other
parse, they can express this using brackets by explicitly writing `true && (false || false)`. This still parses
successfully because the `||` operator is no longer the **direct** child of the
`&&` operator, but is instead the direct child of the `()` operator, and the
`&&` operator is an **indirect** parent.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+first `Bool` non-terminal is not preceded by any terminals, and the last `Bool`
+non-terminal is not followed by any terminals. As a result of this, we apply
+the priority rule to both children of `&&`. However, in the `()` operator, 
+the sole non-terminal is both preceded by and followed by a terminal. As a
```
```suggestion
the sole non-terminal is both preceded by and followed by terminals (`"("` and `")"`, respectively). As a
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+
+Priority blocks will not help us here: the problem comes between two parses
+where both possible parses have a direct parent and child which is within a
+single priority block.
```
```suggestion
Priority blocks will not help us here: the problem comes between two parses
where both possible parses have a direct parent and child which is within a
single priority block (in this case, the `&&` is in the same priority block as itself).
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+with equal priority; a right-associative symbol cannot appear as a direct
+leftmost child of a symbol with equal priority; and a non-associative symbol
+cannot appear as a direct leftmost **or** rightmost child of a symbol with
+equal priority.
```
```suggestion
This is where the notion of associativity comes into play. Associativity applies the following additional rule to parses:

-   a left-associative symbol cannot appear as a direct rightmost child of a symbol
with equal priority,
-   a right-associative symbol cannot appear as a direct leftmost child of a symbol with equal priority, and
-   a non-associative symbol cannot appear as a direct leftmost **or** rightmost child of a symbol with equal priority.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+This means that none of `&&`, `^`, and `||` can appear as the right child of
+any of `&&`, `^`, or `||`. As a result of this, this grammar is also not
+ambiguous. However, it expresses a different grammar, and you are encouraged
+to think about what the differences are in practice.
```
Please provide some example programs here that users can call `kast` on and see the difference, rather than just saying "think about it".

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+associativity relations.
+
+Consider the following grammar, which we will name `lesson-03-h.k` and which
+will express the exact same grammar relation as `lesson-03-f.k`
```
```suggestion
Consider the following grammar, which we will name `lesson-03-h.k` and which
will express the exact same grammar as `lesson-03-f.k`
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+  syntax Exp ::= "true" | "false"
+  syntax Stmt ::= "if" "(" Exp ")" Stmt
+                | "if" "(" Exp ")" Stmt "else" Stmt
+                | ";"
```
```suggestion
                | "{" "}"
```

The empty block `{ }` may read better in the examples below.

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
````diff
@@ -1 +1,627 @@
+```
+
+Roughly, we see that the ambiguity is between an `if` with an `else` or an `if`
+without an `else`. Siknce we want to pick the first parse, we can tell K to
````
```suggestion
without an `else`. Since we want to pick the first parse, we can tell K to
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+at the top of the ambiguous part of the parse, we remove those parses from
+consideration and consider only those remaining. The `prefer` attribute behaves
+similarly, but instead removes all parses which do not have that attribute.
+In both cases, no action is taken if the parse is not ambiguous.
```
```suggestion
Here we have added the `avoid` attribute to the `else` production. Now,
when an ambiguity occurs, any of the possible parses which has the `avoid` symbol
at the top will be removed from consideration. The `prefer` attribute behaves
similarly, but instead removes all parses which do not have a `prefer` symbol at the top.
In both cases, no action is taken if the parse is not ambiguous.
```

dwightguth replied:
It is actually a clearer explanation the way I originally left it. Your wording doesn't adequately explain that what matters is where the ambiguous part of the parse begins.

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
````diff
@@ -1 +1,627 @@
+
+```k
+syntax IntConstant ::= r"(([1-9][0-9]*)|(0[0-7]*)|(0[xX][0-9a-fA-F]+))(([uU][lL]?)|([uU]((ll)|(LL)))|([lL][uU]?)|(((ll)|(LL))[uU]?))?" [token]
+```
````
This is much too overwhelming of an example to start with. Let's begin with just something like `[1-9][0-9]*` or the like, and then show them this one.

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+Here we declare that `main` is a token of sort `Id`. Instead of being parsed
+as a symbol, it gets parsed as a token instead, generating a typed string
+in the AST. This is useful because the parser generally does not treat the
+`main` function in C specially; only the semantics treats it specially.
```
```suggestion
Here we declare that `main` is a token of sort `Id`. Instead of being parsed
as a symbol, it gets parsed as a token, generating a typed string
in the AST. This is useful because the parser generally does not treat the
`main` function in C specially; only the semantics treats it specially.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,627 @@
+performance if you have to repeatedly parse strings with the same parser.
+For this reason, it is generally encouraged that when parsing programs,
+you use K's ahead-of-time parser generation instead. K makes use of
+[GNU Bison](https://www.gnu.org/software/bison/) to generate parsers.
```
```suggestion
So far we have been entirely focused on K's support for just-in-time
parsing, where the parser is generated on the fly prior to being used.
This benefits from being faster to generate the parser, but it suffers in
performance if you have to repeatedly parse strings with the same parser.
For this reason, it is generally encouraged that when parsing programs,
you use K's ahead-of-time parser generation. K makes use of
[GNU Bison](https://www.gnu.org/software/bison/) to generate parsers.
```

ehildenb commented:
I've gone through and provided copy tweaks here and there, but this really needs a larger overhaul, including:

-  Split into at least 2 lessons (it's way too long as is, easy to lose focus). One suggestion would be a single lesson which explains BNF, bracket, and token, and a second lesson which explains priority, assoc, and prefer/avoid.
-  There are many sections which are _only_ explanatory, without offering any real things for the user to try. These section serve better as a reference manual than a tutorial. Every section needs to have something explicit for the user to try at the end.

Commit:
Merge branch 'master' into lesson_3

dwightguth commented:
I'm fine splitting the lesson in half, but I don't understand what you mean in your second bullet point. The general structure of the tutorial is that the lesson explains concepts, and then we provide exercises to synthesize and practice that knowledge at the end. Are you saying you would rather have exercises after each section?

Commit:
copy fixes

Commit:
Merge branch 'master' into lesson_3

Commit:
rename later lessons

Commit:
split lesson into two

Commit:
rename lesson titles

Commit:
fix module names

Commit:
clarify text

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,362 @@
+
+We cannot interpret this program yet, because we have not given rules defining
+the meaning of the `&&` function yet, but we can parse it. To do this, you can
+run (from the same  directory):
```
```suggestion
run (from the same directory):
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
````diff
@@ -1 +1,362 @@
+line. The `--output` flag controls how the resulting AST is represented; don't
+worry about the possible values yet, just use `kore`.
+
+You ought to get the following AST printed on standard output, minus the
+formatting:
+
+```
+inj{SortBoolean{}, SortKItem{}}(
+  Lbl'UndsAnd-And-UndsUnds'LESSON-03-A'Unds'Boolean'Unds'Boolean'Unds'Boolean{}(
+    Lbltrue'Unds'LESSON-03-A'Unds'Boolean{}(),
+    Lblfalse'Unds'LESSON-03-A'Unds'Boolean{}()
+  )
+)
+```
+
+Don't worry about what exactly this means yet, just understand that it
+represents the AST of the program that you just parsed. You ought to be able
+to recognize the basic shape of it by seeing the words `true`, `false`, and
+`And` in there. This is **Kore**, the intermediate representation of K, and we
+will cover it in detail later.
````
I think you should have them do `--output kast` as well, and have them compare the Kore and the KAST, make sure they can map the `_&&__LESSON-03-A_Boolean_Boolean_Boolean` symbol to the munged version. You could just say "Now try `--output kast` and see how they compare" here as something they can try.

ehildenb replied:
Just giving them something to do like this will force them to actually read the output.

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
````diff
@@ -1 +1,362 @@
+true && false || false
+```
+
+When you try and parse this program, you ought to see the following error:
````
```suggestion
Try parsing this program (`kast and-or.bool`), you ought to see the following error:
```

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,362 @@
+unique AST with no error. If you look, you might notice that the bracket itself
+does not appear in the AST. In fact, this is a property unique to brackets:
+productions with the bracket attribute are not represented in the parsed AST
+of a term, and the child of the bracket is folded immediately into the parent
+term. This is the reason for the requirement that a bracket production have
+a single non-terminal of the same sort as the production itself.
```
```suggestion
of a term, and the child of the bracket is folded immediately into the parent
term. This is the reason for the requirement that a bracket production have
a single non-terminal of the same sort as the production itself.

Confirm using `kast` that indeed, the bracket production defined above does
not appear in the resulting output kore on these two examples.
```

dwightguth replied:
This is redundant; we already talk about using kast and noticing that the bracket doesn't appear in the resulting AST.

ehildenb replied:
We talk about it, but this is about giving the user an exercise (so you could put the `### Exercise` block here). I think many people learn better if they are given little tasks to confirm their understanding, so that they have to figure out "what do I need to type in the terminal to confirm that the brackets do not show up?" and "how do I read the output and make sure that the bracket does not show up?"

ehildenb replied:
Ah I see you gave exactly this exercise below. Sounds good!

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,362 @@
+If you have a K definition named `foo.k`, and it generates a directory when
+you run `kompile` called `foo-kompiled`, you can invoke the ahead-of-time
+parser you generated by running `foo-kompiled/parser_PGM <file>` on a file.
+
```
```suggestion

Time the difference in parse time for the program `and-or.bool` above using K's
just-in-time parser (`time kast and-or.bool --output kore`), and the
ahead-of-time parser (`time lesson-03-kompiled/parser_PGM and-or.bool`).
Note that the ahead-of-time parser can _only_ output kore, so no need for the
`--output kore` flag.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/04_disambiguation/README.md:
```diff
@@ -0,0 +1,296 @@
+
+The purpose of this lesson is to teach how to use K's builtin features for
+disambiguation to transform an ambiguous grammar into an unambiguous one that
+expresses the intended the intended ASTs.
```
```suggestion
expresses the intended ASTs.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/04_disambiguation/README.md:
```diff
@@ -0,0 +1,296 @@
+the priority rule to both children of `&&`. However, in the `()` operator, 
+the sole non-terminal is both preceded by and followed by terminals. As a
+result, the priority rule is not applied when `()` is the parent. Because of
+this, the program we mentioned above successfully parses.
```
```suggestion
this, the program we mentioned above successfully parses.

Inspect the kore output by parsing the program `true && true || false`, to make
sure it selects the correct parse. Switch the priority of `&&` and `||` and re-parse,
and inspect that the alternate parse is selected.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/04_disambiguation/README.md:
```diff
@@ -0,0 +1,296 @@
+This means that none of `&&`, `^`, and `||` can appear as the right child of
+any of `&&`, `^`, or `||`. As a result of this, this grammar is also not
+ambiguous. However, it expresses a different grammar, and you are encouraged
+to think about what the differences are in practice.
```
```suggestion
to think about what the differences are in practice.

Try again to parse the program `true && true || false`, and then the program
`true || true && false` using the left-associative grammar in `LESSON-04-C`.
Inspect the kore output for each, to understand how it selects each parse.
You can also try the same with `left` replaced by `right` in the grammar.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/04_disambiguation/README.md:
```diff
@@ -0,0 +1,296 @@
+  syntax Boolean ::= "!" Boolean [not, function]
+  syntax Boolean ::= Boolean "&&" Boolean [and, function]
+  syntax Boolean ::= Boolean "^" Boolean [xor, function]
+  syntax Boolean ::= Boolean "|" Boolean [or, function]
```
```suggestion
  syntax Boolean ::= "true"               [literal]
                   | "false"              [literal]
                   | "(" Boolean ")"      [atom, bracket]
                   | "!" Boolean          [not, function]
                   | Boolean "&&" Boolean [and, function]
                   | Boolean "^" Boolean  [xor, function]
                   | Boolean "|" Boolean  [or, function]
```

Might as well get them in the habit of seeing/writing more legible K early.

ehildenb commented on k-distribution/k-tutorial/1_basic/04_disambiguation/README.md:
```diff
@@ -0,0 +1,296 @@
+at the top of the ambiguous part of the parse, we remove those parses from
+consideration and consider only those remaining. The `prefer` attribute behaves
+similarly, but instead removes all parses which do not have that attribute.
+In both cases, no action is taken if the parse is not ambiguous.
```
```suggestion
In both cases, no action is taken if the parse is not ambiguous.

Parse the program `if (true) {} if (false) {} else {}` using the grammar of
`LESSON-04-F`, to see how it builds the kore AST. Now place the `avoid`
attribute on the other `"if"` production instead, and see how this changes the
kore AST. Alternatively, just swap in `prefer` for the existing `avoid`.
```

Commit:
fix some of the comments

dwightguth commented:
I addressed some of your comments here. It seems a number of them seem to be trying to insert miniature exercises into the body of the lesson, though. This is not the (current) structure of the lesson, and it breaks up the flow considerably, not to mention it's a considerable deviation in the narrative style of the text. It's also an anomaly when considering the structure of previous lessons. I'm not necessarily opposed to having this type of miniature exercise in the text (although I wonder if we are overdoing things), but we need to come up with a much better way of embedding it into the document if we are going to.

Commit:
fix comments regarding mid-lesson exercises

ehildenb commented on k-distribution/k-tutorial/1_basic/03_parsing/README.md:
```diff
@@ -1 +1,397 @@
+
+### Exercise
+
+Parse the expression `false || true` with `--output kastr`. See if you can
```
```suggestion
Parse the expression `false || true` with `--output kast`. See if you can
```

ehildenb commented on k-distribution/k-tutorial/1_basic/04_disambiguation/README.md:
```diff
@@ -0,0 +1,312 @@
+
+Parse the program `true && false || false` yourself, and confirm that the AST
+places `||` as the top level symbol. Then modify the definition so that you
+will get the opposite parse.
```
```suggestion
will get the alternative parse.
```

ehildenb commented on k-distribution/k-tutorial/1_basic/04_disambiguation/README.md:
```diff
@@ -0,0 +1,312 @@
+
+Parse the program `true && false && false` yourself, and confirm that the AST
+places the rightmost `&&` at the top of the expression. Then modify the 
+definition to generate the opposite parse.
```
```suggestion
definition to generate the alternative parse.
```

Commit:
Apply suggestions from code review

Co-authored-by: Everett Hildenbrandt <everett.hildenbrandt@gmail.com>

Commit:
fix one more piece of wording

rv-jenkins committed:
Merge branch 'master' into lesson_3
